src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -3065,6 +3066,9 @@ private void SubstituteFakeTokens(Blob blob, ImmutableArray<byte> methodBodyIL,
                                         case LiteralGreatestMethodDefinitionToken:
                                             token = GreatestMethodDefIndex;
                                             break;
+                                        case SourceDocumentIndex:
+                                            token = _dynamicAnalysisDataWriterOpt.GetOrAddDocument(((CommonPEModuleBuilder)module).GetSourceDocumentFromFakeToken((uint)(pseudoToken & 0x00ffffff)));
```

Comment:
Poking around a little, I'm not sure that there's a natural thing to add to IModule (which is a CCI interface) which solves the issue.

It turns out that all IModule instances in Roslyn are CommonPEModuleBuilder instances, so the cast is sound. We could change the type of the module member of MetadataWriter to be typed as CommonPEModuleBuilder, but that would require lots of editing.